h2o-core/src/main/java/water/util/MathUtils.java

```diff
@@ -377,17 +377,24 @@ public static boolean equalsWithinRecSumErr(double actual, double expected, int
     return Math.abs(actual - expected) <= (n-1) * Math.ulp(actual) * absum;
   }
 
-  /** Compare 2 doubles within a tolerance
+  /** 
+   * Compare 2 doubles within a tolerance. True iff the numbers difference within a absolute tolerance or
+   * within an relative tolerance.
+   * 
    *  @param a double
    *  @param b double
-   *  @param abseps - Absolute allowed tolerance
-   *  @param releps - Relative allowed tolerance
+   *  @param absoluteTolerance - Absolute allowed tolerance
+   *  @param relativeTolerance - Relative allowed tolerance
    *  @return true if equal within tolerances  */
-  public static boolean compare(double a, double b, double abseps, double releps) {
-    return
-      Double.compare(a, b) == 0 || // check for equality
-      Math.abs(a-b)/Math.max(a,b) < releps ||  // check for small relative error
-      Math.abs(a - b) <= abseps; // check for small absolute error
+  public static boolean compare(double a, double b, double absoluteTolerance, double relativeTolerance) {
+    assert absoluteTolerance >= 0;
+    assert relativeTolerance >= 0;
+    
+    final boolean equal = Double.compare(a, b) == 0;
```

Comment:
This deoptimization missed the point of the exact comparison, the exact equality is fast-path optimization to avoid division. By removing the lazy evaluation you are making this comparison redundant, it will be caught either by relative or absolute error check later.

Reply:
Maybe it would be a good idea to set this repository to automatically cancel all older reviewes once new changes are pushed. I originally reviewed a much smaller change: https://github.com/h2oai/h2o-3/commit/434467971495454b4f74c21284bb718f283e5212

Reply:
Anyway I have reordered it

Reply:
@Pscheidl I try to rerequest :+1: when I do some bigger changes but I forget sometimes :-/ The original commit was mainly my expression of surprise I run into this issue with some minimal tests to show it.

Reply:
That was a generic comment.